### PR TITLE
[refactor] Improve `Function` and `ImportModule`

### PR DESCRIPTION
**Pull request by apepkuss**
In this PR, the following improvements are introduced:

- In `wasmedge-sys`, improve the design of `Function` and `ImportModule`.
- Correspondingly, `Func`, `ImportObjectBuilder`, `VmBuilder` and `Store` in `wasmedge-sdk` are refactored.

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Review by L-jasmine — APPROVED** [review]

File: crates/wasmedge-sys/src/instance/function.rs [hunk 1]
```diff
@@ -380,11 +380,11 @@ impl Function {
     pub fn create_sync_func<T>(
         ty: &FuncType,
         real_fn: BoxedFn,
-        data: Option<&mut T>,
+        data: Option<Box<T>>,
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply crates/wasmedge-sys/src/instance/function.rs:383]: We'll consider it in the coming PR. Thanks a lot!

File: crates/wasmedge-sys/src/instance/function.rs [hunk 2]
```diff
@@ -463,11 +463,17 @@ impl Function {
     /// * If fail to create a [Function], then [WasmEdgeError::Func(FuncError::Create)](crate::error::FuncError) is returned.
     ///
     #[cfg(all(feature = "async", target_os = "linux"))]
-    pub fn create_async_func(
+    pub fn create_async_func<T: Send + Sync>(
         ty: &FuncType,
         real_fn: BoxedAsyncFn,
+        data: Option<Box<T>>,
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply crates/wasmedge-sys/src/instance/function.rs:469]: We'll consider it in the coming PR. Thanks a lot!

File: crates/wasmedge-sys/src/instance/module.rs [hunk 3]
```diff
@@ -388,14 +388,13 @@ pub trait AsInstance {
 
 /// An [ImportModule] represents a host module with a name. A host module consists of one or more host [function](crate::Function), [table](crate::Table), [memory](crate::Memory), and [global](crate::Global) instances,  which are defined outside wasm modules and fed into wasm modules as imports.
 #[derive(Debug, Clone)]
-pub struct ImportModule<T: Send + Sync + Clone> {
+pub struct ImportModule {
     pub(crate) inner: Arc<InnerInstance>,
     pub(crate) registered: bool,
     name: String,
-    _host_data: Option<Box<T>>,
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply crates/wasmedge-sys/src/instance/module.rs:395]: We'll consider it in the coming PR. Thanks a lot!

File: src/externals/function.rs [hunk 4]
```diff
@@ -119,16 +119,18 @@ impl Func {
             > + Send
             + Sync
             + 'static,
+        data: Option<Box<T>>,
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply src/externals/function.rs:122]: We'll consider it in the coming PR. Thanks a lot!

File: src/externals/function.rs [hunk 5]
```diff
@@ -44,7 +44,7 @@ impl Func {
             + Send
             + Sync
             + 'static,
-        data: Option<&mut T>,
+        data: Option<Box<T>>,
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply src/externals/function.rs:47]: We'll consider it in the coming PR. Thanks a lot!

File: src/import.rs [hunk 6]
```diff
@@ -134,15 +132,18 @@ impl<T: Send + Sync + Clone> ImportObjectBuilder<T> {
             > + Send
             + Sync
             + 'static,
+        data: Option<Box<D>>,
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply src/import.rs:135]: We'll consider it in the coming PR. Thanks a lot!

File: src/plugin.rs [hunk 7]
```diff
@@ -268,7 +268,7 @@ impl<T: Send + Sync + Clone> PluginModuleBuilder<T> {
             + Send
             + Sync
             + 'static,
-        data: Option<&mut D>,
+        data: Option<Box<D>>,
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply src/plugin.rs:271]: We'll consider it in the coming PR. Thanks a lot!

File: src/plugin.rs [hunk 8]
```diff
@@ -309,15 +309,18 @@ impl<T: Send + Sync + Clone> PluginModuleBuilder<T> {
             > + Send
             + Sync
             + 'static,
+        data: Option<Box<D>>,
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply src/plugin.rs:312]: We'll consider it in the coming PR. Thanks a lot!

**Comment by apepkuss** [comment]
@L-jasmine Thanks for the review!

**Pull request merged**